src/alpaka/AlpakaCore/ScopedContext.h

```diff
@@ -121,12 +121,10 @@ namespace cms::alpakatools {
 
       template <typename TQueue>
       void enqueueCallback(TQueue& stream) {
-        alpaka::enqueue(stream, [holder = waitingTaskHolder_]() {
-          // TODO: The functor is required to be const, so can't use
-          // 'mutable', so I'm copying the object as a workaround. I
-          // wonder if there are any wider implications.
-          auto h = holder;
-          h.doneWaiting(nullptr);
+        alpaka::enqueue(stream, [holder = std::move(waitingTaskHolder_)]() {
+          // The functor is required to be const, but the original waitingTaskHolder_
+          // needs to be notified...
+          const_cast<edm::WaitingTaskWithArenaHolder&>(holder).doneWaiting(nullptr);
```

Comment:
@makortel can you check this part ?

Without it, using a non-blocking queue on the CPU would wait forever; out interpretation is that the `waitingTaskHolder_` was not being properly notified.

Reply:
Reminding myself of the exact behavior of `WaitingTaskWithArenaHolder`, the `const_cast` is the correct action. The `holder` is reference counted, and leaving the refcount of the object held by the lambda decreased by the destructor (instead of the explicit `doneWaiting()` as here) would defer the "notification" until that lambda is destructed (whenever that is).

I wonder if `alpaka::enqueue()` strictly requires non-mutable functors. On the other hand, if `alpaka::enqueue()` does not propagate errors from the asynchronous work (I don't know), we probably want to do the `WaitingTaskWithArenaHolder` notification differently anyway.

Reply:
Thanks !

About

> I wonder if `alpaka::enqueue()` strictly requires non-mutable functors.

I didn't experiment much with it, I only tested that without the `const_cast` it doesn't seem to compile.